triton-constraint-circuit/src/lib.rs

```diff
@@ -1471,4 +1705,316 @@ mod tests {
+    /// Therefore, in order to sample a random satisfying input to `C" || k`, it
+    /// suffices to select `input` at random, define `C"(X) = C(X) - C(input)`,
+    /// and evaluate `some_expr(input)`. Then `input || some_expr(input)` is a
+    /// random satisfying input to `C" || k`. It follows** that `input` is a
+    /// satisfying input to `C"`.
+    ///
+    /// (**: If the transformation is sound.)
+    ///
+    /// This description makes use of the following commutative diagram.
+    ///
+    /// ```notest
+    ///          C ---- degree-lowering -----> C || k
+    ///          |                               |
+    /// subtract |                      subtract |
+    ///    fixed |                         fixed |
+    ///   output |                        output |
+    ///          |                               |
+    ///          v                               v
+    ///          C* --- degree-lowering ----> C* || k
+    /// ```
+    ///
+    /// The point of this elaboration is that in this particular case, testing completeness
+    /// and soundness require the same code path. If `input` and `input || some_expr(input)`
+    /// work for circuits before and after degree lowering, this fact establishes
+    /// both completeness and soundness simultaneously.
+    ///
+    /// Shrinking on this test is disabled because we noticed some weird ass behavior.
+    /// In short, shrinking does not play ball with the arbitrary circuit generator;
+    /// it seems to make the generated circuits *more* complex, not less so.
+    #[proptest(cases = 1000, max_shrink_iters = 0)]
+    fn node_substitution_is_complete_and_sound(
+        #[strategy(arbitrary_circuit_monad(10, 10, 10, 200, 10))] mut multicircuit_monad: Vec<
+            ConstraintCircuitMonad<SingleRowIndicator>,
+        >,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_main_inputs(&#multicircuit_monad)))]
+        #[filter(!#main_input.is_empty())]
+        main_input: Vec<BFieldElement>,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_aux_inputs(&#multicircuit_monad)))]
+        #[filter(!#aux_input.is_empty())]
+        aux_input: Vec<XFieldElement>,
+        #[strategy(vec(arb(), ConstraintCircuitMonad::num_challenges(&#multicircuit_monad)))]
+        challenges: Vec<XFieldElement>,
+        #[strategy(arb())] substitution_node_index: usize,
+    ) {
+        let mut main_input = Array2::from_shape_vec((1, main_input.len()), main_input).unwrap();
+        let mut aux_input = Array2::from_shape_vec((1, aux_input.len()), aux_input).unwrap();
+
+        let output_before_lowering = multicircuit_monad
+            .iter()
+            .map(|constraint| {
+                constraint.circuit.borrow().evaluate(
+                    main_input.view(),
+                    aux_input.view(),
+                    &challenges,
+                )
+            })
+            .collect_vec();
```

Comment:
I find this easier to follow:

```suggestion
        let output_before_lowering = multicircuit_monad
            .iter()
            .map(|m| m.circuit.borrow())
            .map(|c| c.evaluate(main_input.view(), aux_input.view(), &challenges))
            .collect_vec();
```

Happy about more descriptive variable names than `m` and `c`.